api/src/main/java/org/eclipse/microprofile/config/inject/ConfigProperties.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.eclipse.microprofile.config.inject;
+
+
+import static java.lang.annotation.ElementType.TYPE;
+import static java.lang.annotation.RetentionPolicy.RUNTIME;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Stereotype;
+
+
+/**
+ * <p>
+ * Retrieve a number of related configuration properties to a property class.
+ * <pre>
+ * &#064;ConfigProperties(prefix="server")
+ * public class MyServer {
```

Comment:
Why is this a class? To retain the same developer experience across all MP, we should align this with MP REST Client. In the REST client, the mechanism is very different. Instead of injecting values into class fields, REST Client creates an implementation of an user interface and then injects it into a bean.

Example in REST client:

```
@RegisterRestClient
public interface MyClient {
}
```

```
@Dependent
public class MyCaller {
  @Inject @RestClient MyClient client;
}
```

Configuration could work similarly:

```
@RegisterConfigProperties(defaultPrefix = "server")
public interface MyConfig {
    String getHost();
    int getPort();
}
```
Then inject using the `@ConfigProperties` annotation:

```
@Dependent
public class MyBean {
   @Inject @ConfigProperties(prefix="server2")  // if prefix not provided, use defaultPrefix from `@RegisterConfigProperties` on the interface
   MyConfig myConfig; 
}
```

Or programmatically:

```
MyConfig myConfig = Config.getValues("server2", MyConfig.class);
```

Reply:
Alternatively, we could support injecting directly to fields of CDI beans as in your current proposal, but I suggest to make it more standard and use `@Inject @ConfigProperty` with every field. Then it would look like this:

```
@ConfigProperties(prefix="server")
public class MyServer {
    @Inject @ConfigProperty public String host; //maps the property name server.host
    @Inject @ConfigProperty(name = "portnumber") public int port;    //maps to the property name server.portnumber
    @Inject @ConfigProperties(prefix="address") Address address //maps the properties with prefix "server.address." to a custom interface Address
}
```

with `Address` being 
```
@RegisterConfigProperties
public interface Address {
    String gethost();
    String getPort();
}
```

Reply:
> @RegisterConfigProperties
public interface Address {
    String gethost();
    String getPort();
}

I don't quite understand what you meant here. For your example, which property names do you mean to map to?
In my example, I want to demonstrate any types can be specified as long as there is a corresponding converter.

Reply:
> @Inject @ConfigProperty public String host; //maps the property name server.host

This is unnecessary and purely boilerplate.

Reply:
By the way, I think my proposal is better than the Rest Client alternative. I am trying to define an entity, so a class with fields match better logically.

Reply:
Yes there are use cases for both classes and interfaces. I understand the motivation behind trying to define and entity. On the other hand, it is unlikely that consumers have these custom entity that match their configuration, so once the feature is out, they would need to create a new object to represent that structure. Do we want to be an interface or a class?

@OndroMih has a very valid point about REST Client using interfaces. For consistency across the platform, we should probably use interfaces as well.

From a developer experience perspective, the API's are very similar, you provide a structure and what MP to instantiate it for you, but one is a class and the other is an interface, so you have to know which one to use. I think it would be great if a single pattern was used, so you always know that for these cases across the platform you always use a structure of type `X`.

Reply:
This is different from MP Rest Client as there is no field injection happening in MP Rest Client. They have very different synmatics. I disagree with going for interface. For doing interface, you will need to establish the JavaBean naming convention. Spring uses [classes](https://www.baeldung.com/configuration-properties-in-spring-boot) for ConfigProperties.